### PR TITLE
Scheduler reconciliation fixes and testing

### PR DESCRIPTION
**Pull request by severinson**

### CODE REVIEW

**Review by zuqq — APPROVED** [review]

File: internal/scheduler/scheduler_test.go [hunk 1]
```diff
@@ -114,11 +136,12 @@ var queuedJob = testfixtures.JobDb.NewJob(
 	uint32(10),
 	schedulingInfo,
 	true,
-	1,
+	0,
```
[the remark this answers is not in the document]
Comment (severinson) [reply internal/scheduler/scheduler_test.go:139]: Yep

File: internal/scheduler/jobdb/job.go [hunk 2]
```diff
@@ -63,6 +69,194 @@ func EmptyJob(id string) *Job {
 	return &Job{id: id, runsById: map[uuid.UUID]*JobRun{}}
 }
 
+func (job *Job) String() string {
+	// Include field names in string representation by default.
+	return fmt.Sprintf("%#v", job)
```
[the remark this answers is not in the document]
Comment (severinson) [reply internal/scheduler/jobdb/job.go:74]: Changing to `%+v` would result in a recursive call I believe.

File: internal/scheduler/scheduler.go [hunk 3]
```diff
@@ -192,15 +197,37 @@ func (s *Scheduler) Run(ctx *armadacontext.Context) error {
...
-// Otherwise, we only generate events from jobs updated since the last cycle.
+// cycle runs one iteration of the scheduling loop.
+//
+// Each iteration of the scheduler is functionally equivalent to the following:
+//  1. Load the entire scheduler state, consisting of jobs, runs, and errors, from the persistent schedulerDb.
+//  2. Generate job state transitions, which we can be seen as performing a mapping
+//     (job_0, runs_0) -> (job_0', runs_0')
+//     ...
+//     (job_n, runs_n) -> (job_n', runs_n'),
+//     where runs_i is the set of jobs associated with job_i, and n is the number of jobs.
```
[the remark this answers is not in the document]
Comment (severinson) [reply internal/scheduler/scheduler.go:208]: Sorted in follow-up pr.

File: internal/scheduler/scheduler.go [hunk 4]
```diff
@@ -192,15 +197,37 @@ func (s *Scheduler) Run(ctx *armadacontext.Context) error {
...
+//     - Triggered by a scheduling decision, e.g., a scheduled job should have an additional run associated with it.
+//  3. Generate eventSequences encoding these state transition and publish these to Pulsar.
+//  4. Wait for these Pulsar eventSequences to be persisted to the schedulerDb.
+//
+// For performance reasons, the actual cycle differs from the above in three ways:
+//   - In each cycle, we only load jobs, runs, and errors that have changed since the last cycle.
+//     For unchanged jobs, runs, and errors, we rely on an in-memory cache maintained between cycles, i.e., the jobDb.
+//   - Similarly, we only perform job state transitions for jobs with changed jobs and runs.
+//     This is unless updateAll is true, in which case we perform state transitions for all jobs.
+//     This is necessary for the first cycle after a leader failOver.
```
[the remark this answers is not in the document]
Comment (severinson) [reply internal/scheduler/scheduler.go:220]: Sorted in follow-up pr.

File: internal/scheduler/scheduler.go [hunk 5]
```diff
@@ -227,7 +249,28 @@ func (s *Scheduler) cycle(ctx *armadacontext.Context, updateAll bool, leaderToke
 		updatedJobs = txn.GetAll()
 	}
 
-	// Generate any events that came out of synchronising the db state.
+	// Load error messages for any failed runs.
+	// TODO(albin): An unbounded number of job runs may fail between subsequent cycles.
+	//              E.g., if 1M runs fail and each one generates a 1Mb error message, we'd need to load 1Tb of errors.
```
[the remark this answers is not in the document]
Comment (severinson) [reply internal/scheduler/scheduler.go:254]: I added a reference to this link. Nice find

**Pull request merged**